### ISSUE TITLE
FFT Backend documentation

### ISSUE THREAD

**Issue by masip85** [scipy/scipy#12509]
**Is your feature request related to a problem? Please describe.**

I am looking for documentation about the fft backends. It is very short. But I found this in a RalfGommers slide share:
![imagen](https://user-images.githubusercontent.com/12863854/86888363-53e38c80-c0fa-11ea-9ea6-657fc1cda7b6.png)

My question here is, what other backends can be tested? According to docs, _PyFFTW provides a way to replace a number of functions in scipy.fft with its own functions, which are usually significantly faster, via pyfftw.interfaces. Because PyFFTW relies on the GPL-licensed FFTW it cannot be included in SciPy. Users for whom the speed of FFT routines is critical should consider installing PyFFTW._

So following release notes and GitHub SciPy and NumPy issues, you both have adopted pocketfft by default. With backend I can use pyfftw with its licenses limitations. But what else can I do with that backend properties? I really don't understand which strings I can use, or which are the kind of objects that are commented here:

    Parameters

    backend: {object, ‘scipy’}

        The backend to use. Can either be a str containing the name of a known backend {‘scipy’}, or an object that implements the uarray protocol.

**Describe the solution you'd like**
Have documentation about which backend can be tested. Also a text which let noobs like me understand which kind of objects are compatible. 

**Describe alternatives you've considered**
A great point is that the list could have an attached capture of the benchmarks (adding build configuration of fftw). Using something like https://github.com/numpy/numpy/pull/11888

**Comment by rgommers** [comment]
Linking to other existing backends in the docs would be good. At least CuPy has one, and maybe mkl-fft (@peterbell10?).

**Comment by peterbell10** [comment]
The full list is:
- PyFFTW -> `from pyfftw.interfaces import scipy_fft`
- MKL FFT -> `from mkl_fft import _scipy_fft_backend`
- CuPy -> `from cupyx.scipy import fft`

Only `"scipy"` can be used as a string. There was discussion of adding other string options but I think that for licensing reasons that wasn't feasible.

**Comment by mreineck** [comment]
While this is being overhauled, it might be worthwhile to re-evaluate the above mentioned paragraph about FFTW. At least without FFTW_MEASURE, I'm pretty confident that pyfftw doesn't outperform `scipy.fft` by much any more.

**Comment by peterbell10** [comment]
That comparison with `FFTW` is in the `pyfftw` docs.

**Comment by mreineck** [comment]
Ah, sorry, I hadn't realized that!

**Comment by masip85** [comment]
@peterbell10 , here? https://pyfftw.readthedocs.io/en/latest/ 
@mreineck 
>I'm pretty confident that pyfftw doesn't outperform scipy.fft by much any more.

I am trying to use neon support in a RPI arm cortex-a72 (_A program linking to an FFTW library compiled with SIMD support can obtain a nonnegligible speedup for most complex and r2c/c2r transforms. In order to obtain this speedup, however, the arrays of complex (or real) data passed to FFTW must be specially aligned in memory (typically 16-byte aligned), and often this alignment is more stringent than that provided by the usual malloc (etc.) allocation routines._) to get these speeds:
http://www.vesperix.com/arm/fftw-arm/bench/gcc-a8-nofma/index.html
So, fftw/pyfftw is not faster, maybe, in normal conditions. But.. there are multiple configurations..

**Comment by mreineck** [comment]
You are right, vectorization can help a lot, both FFTW and `scipy.fft`. Unfortunately there are two problems with this on the ARM target:
- `scipy` binary packages don't seem to be optimized for different CPUs of a given architecture, just for the "lowest common denominator". For example, the binary package for x86_64 only supports SSE2 vectorization, and `scipy.fft` therefore cannot use any longer vectors. Most x86 CPUs have AVX instructions these days, and `scipy.fft` could run much faster with these, but there is no way to use this unless you compile your own `scipy` version and fiddle with the compiler flags.
- even if the issue above was solved, I don't know enough about ARM vector instructions to support the properly in `scipy.fft`. If you can tell me which preprocessor macros I have to check just to determine how long vector registers for `float` and `double` are, that would already be a big step forward.

On x86, I currently do this (slightly simplified):

```
#if defined(__AVX512F__)
constexpr size_t vsize_float=16;
constexpr size_t vsize_double=8;
#elif defined(__AVX__)
constexpr size_t vsize_float=8;
constexpr size_t vsize_double=4;
#else
constexpr size_t vsize_float=4;
constexpr size_t vsize_double=2;
#endif 
```

That's really all I need; the compiler will take care of the rest.

(Array alignment etc. is taken care of by the library as well)

**Comment by masip85** [comment]
@mreineck , I don't know if I have understood well, but I am always talking about using scipy with a different backend. In this case, I am trying to wrap fftw compiled with neon with pyfft. And with pyfftw interface no changes should be needed, isn't it? Maybe I am making mistakes in my reasoning.

**Comment by mreineck** [comment]
You are right that this scenario will need no changes. I was just saying that if I could get some help with tuning `scipy.fft` for ARM architectures, it might be pretty close to FFTW in performance on that hardware, too. If at all possible, I'd like to make that happen :)

**Comment by masip85** [comment]
Well, that is the reason scipy.fft.backend exists,isn't it? I can help you,but hardly. I have this pipeline, but I need a little bit of translation.

**Comment by mreineck** [comment]
Thank you! I'm trying to revive #10463 which was opened exactly for this purpose. If there is anything I can clarify, please let me know!

**Comment by mreineck** [comment]
Coming back to this:

> That comparison with FFTW is in the pyfftw docs.

Hmm, it's also in https://docs.scipy.org/doc/scipy/reference/tutorial/fft.html, which seems to be generated from
`doc/source/tutorial/fft.rst` in the `scipy` repo. Any opinions on keeping or removing this note? I don't feel strongly either way, but having this note directly at the top of `scipy`'s FFT reference docs feels a bit defensive ...

**Comment by gabrielfougeron** [comment]
> The full list is:
> 
> * PyFFTW -> `from pyfftw.interfaces import scipy_fft`
> * MKL FFT -> `from mkl_fft import _scipy_fft_backend`
> * CuPy -> `from cupyx.scipy import fft`
> 
> Only `"scipy"` can be used as a string. There was discussion of adding other string options but I think that for licensing reasons that wasn't feasible.

This was very helpful to me, I wish it made it in the official docs

**Comment by rgommers** [comment]
Given that we plan to deprecate/remove Uarray (xref gh-25878), I'll close this issue - doesn't seem worth spending time on now.
